### PR TITLE
ScanVolumes: Do not filter out multiple ESP partitions

### PR DESCRIPTION
**Pull request by bcully**
Honestly, I'm not sure exactly why we're deduplicating here at all, but deduplicating by VolUuid makes it impossible to use bootloaders on multiple EFI partitions, at least of the standard type.

### CODE REVIEW

**Comment by dakanji** [comment]
Hi,

Would typically have asked you to push this upstream as I don't know the ins and outs of the matter of this old segment of code but seems the upstream development is temporarily stalled and you are not likely to get a response.

So, it appears this was added torEFInd for this reason back in 2014:
```
- Scans of ext2/3/4fs and ReiserFS partitions now omit partitions with
  duplicate filesystem UUIDs. These are likely parts of RAID arrays and so
  would have the same boot loaders or kernels as the first one with a given
  UUID.
```

Question is then on your proposed changes impact this reasoning for the code. That is, what happens with such RAID Arrays with a switch to PartGUID?

**Comment by bcully** [comment]
Thanks for looking into this. It's an interesting point. I guess that the GPT partition IDs of each RAID replica would indeed be different on each mirror, and so scans could produce duplicate entries. One thing I would note is that the comment refers to ext and reiser partitions, where we would expect a unique partition ID for every fs creation. What I am getting bitten by is that it is also applying to EFI partitions, where they will tend to have the same partition ID every time (unless you create the ESP by hand instead of letting your OS installer do it, I suppose).

So perhaps we could deduplicate by partition ID *unless the partition is an EFI system partition*? I haven't looked yet but I imagine we could filter by partition type or something...

**Comment by bcully** [comment]
I've just updated with the alternative I proposed, where we don't try to deduplicate ESP partitions. There's an even more conservative approach possible, where we just ignore the `IsReadable` bit when looking for volumes specified in manual stanzas, but I think the approach I've taken here is friendlier for users that just want scan to show them all bootable loaders. Just for comparison, both Clover and OpenCore find the bootloaders on my other ESP partitions already.

**Review by joevt — CHANGES_REQUESTED** [review]
Before this can be approved, a few things need to be known:

1) Where does VolUuid come from on different types of partitions (FAT, HFS+, NTFS, Ext4) and disks (GPT, MBR)? I think it's a volume ID stored in the file system? Usually the first block of a file system has this number.
`sudo dd if=/dev/rdisk0s1 count=1 | xxd`

2) Where does PartGuid come from on different types of partitions (FAT, HFS+, NTFS, Ext4) and disks (GPT, MBR)? I think this is a partition ID stored in the GPT? So what does MBR use in this case? Does this patch work for MBR disks which doesn't have partition GUIDs?

Maybe the patch needs to compare both? But that won't be sufficient for two MBR disks that have an EFI partition as the first partition in the same location. In that case you would also have to add a disk ID (I think MBR has a disk ID and GPT has a disk UUID).

But adding Disk ID won't be sufficient if the same disk can exist under different EFI device paths. We may want to be able to boot the same disk using different EFI device paths, so maybe we want to compare device paths? Then we wouldn't need to check partition and volume and disk IDs...

I suppose it would be helpful for the debug output to print these identifiers.

**Comment by bcully** [comment]
I think the approach I've taken in the latest update is less disruptive, making some of these questions less critical. I'm no longer switching from comparint VolUuids to PartGuids, I'm just skipping the duplicate check if the partition type is ESP. For all other partitions types, the logic is unchanged.

I believe you're right about VolUuid, it comes from the file system header. I also think that GPT partition UUIDs are generated by the partitioning tool before any file system is placed into the partitions.

I don't know how EFI constructs the partition type GUID for MBR partition tables. I think it should be alright though, because we have two cases to deal with:

1. The EFI partition type for the ESP is the same on MBR and GPT tables. Great, we'll scan all the EFI partitions
2. The EFI partition type on MBR isn't the same as on GPT. In that case, we fall through to the status quo, filtering out subsequent EFI partitions. I don't think this is what we want, but it isn't a regression. If someone with multiple MBR disks with EFI partitions on them is around to test, we can add a second conditional.

> Before this can be approved, a few things need to be known:
> 
>     1. Where does VolUuid come from on different types of partitions (FAT, HFS+, NTFS, Ext4) and disks (GPT, MBR)? I think it's a volume ID stored in the file system? Usually the first block of a file system has this number.
>        `sudo dd if=/dev/rdisk0s1 count=1 | xxd`
> 
>     2. Where does PartGuid come from on different types of partitions (FAT, HFS+, NTFS, Ext4) and disks (GPT, MBR)? I think this is a partition ID stored in the GPT? So what does MBR use in this case? Does this patch work for MBR disks which doesn't have partition GUIDs?
> 
> 
> Maybe the patch needs to compare both? But that won't be sufficient for two MBR disks that have an EFI partition as the first partition in the same location. In that case you would also have to add a disk ID (I think MBR has a disk ID and GPT has a disk UUID).
> 
> But adding Disk ID won't be sufficient if the same disk can exist under different EFI device paths. We may want to be able to boot the same disk using different EFI device paths, so maybe we want to compare device paths? Then we wouldn't need to check partition and volume and disk IDs...
> 
> I suppose it would be helpful for the debug output to print these identifiers.

**Comment by bcully** [comment]
I've just noticed that you left suggested changes, which I can no longer see since I overwrote the branch. I apologize for clobbering them, please feel free to apply them again and I'll try not to overwrite the branch without looking next time.

**Comment by joevt** [comment]
I'm ok with not removing duplicate EFI partitions, if they are not the same EFI partition.
MBR disks shouldn't have an EFI partition, only GPT disks should. Macs can have hybrid MBR/GPT but I believe only the GPT is used in EFI. The MBR is used to boot Windows in Legacy BIOS mode.

**Comment by joevt** [comment]
I didn't make any change suggestions. Maybe I clicked the wrong comment option.

**Comment by dakanji** [comment]
@bcully ... Can you make an upstream merge request for the EFI Partition filter?

**Comment by dakanji** [comment]
Sorry ... Cancelled the merge as I am now seeing duplication of my EFI partition.

**Comment by dakanji** [comment]
Putting this on hold for now.

I will implement parts of the PR manually and will later implement the filter as you proposed but will tie this to a user selectable option in the config.

**Comment by dakanji** [comment]
New config option, `allow_duplicates` has been introduced.
Activate this to get the same effect (See [config.conf-sample](https://github.com/dakanji/RefindPlus/blob/GOPFix/config.conf-sample) file)

**Comment by bcully** [comment]
> Sorry ... Cancelled the merge as I am now seeing duplication of my EFI partition.

Out of curiosity, what do you mean by duplicate here? Are you RAID1ing your EFI partition, is EFI creating multiple volumes for the same partition, or something else? It'd be nice to capture the issue in case there's a way to handle it more intelligently...

**Comment by dakanji** [comment]
Understand but happier with this lighter touch at this time. Allows things to work as they always have as conscious that this has been sitting there for ages without many seeming to have problems with it. 

Anyone that need different just needs to change a setting. 

Basically settling for "good enough" compromise on this.

**Comment by joevt** [comment]
Maybe comparing device paths is the way to go? Do your duplicates that you want to remove have the same device path, while the duplicates that you don't want to remove have different device paths?

**Comment by dakanji** [comment]
@bcully ... it was a quirk with my setup that gave me the result.
I have an additional fallback bootloader present that does not normally shown up but now does with this setup and your last changes would work well in normal setups.

Will figure out a way to accommodate this both.

**Comment by startergo** [comment]
> @bcully ... It was a quirk with my setup that gave me the result and have a better understanding of the matter.
> 
> I have an additional fallback bootloader present that does not normally shown up but now does with this setup.
> Actually, the conventional setup with rEFInd is not to show bootloaders in other ESPs. The proposal would have changed this default behaviour.
> 
> The conventional way to access such is to use manual stanzas. Having a config option as now provided here allows for overriding this behaviour but should not be the default.

rEFInd 0.11.6 and later find EFI loaders on other partitions. For instance OpenCore is on the internal SSD's EFI and rEFInd is on an USB drive. rEFInd finds both OpenCore and the bootprotect loader.

**Comment by dakanji** [comment]
> rEFInd 0.11.6 and later find EFI loaders on other partitions

No change to rEFInd 0.12.0 code and it does not on mine and apparently also not on the OP's as well which is what generated this in the first place.

Not sure what the the target is at this point since an option has been provided for those that want this regardless.

**Comment by dakanji** [comment]
> rEFInd is on an USB 

Just noticed this. 

It will of find OpenCore on an EFI portion since rEFInd iitself is not on one. I.E., OpenCore I EFI is the only EFI being scanned.

If you put rEFInd in an EFI partition, then your OpenCore will no longer show up.

Alternatively, if you create a second EFI partition with a loader in it, only one of them will show up.

**Comment by startergo** [comment]
I meant rEFInd is on USB inside the ESP

**Comment by dakanji** [comment]
Don't understand what that means. 
Can you perhaps move it to a discussion if you want to discuss further instead?

**Comment by startergo** [comment]
> Don't understand what that means. 
> Can you perhaps move it to a discussion if you want to discuss further instead?

You said 
> If you put rEFInd in an EFI partition, then your OpenCore will no longer show up.

I said my rEFInd was in an EFI partition on a USB stick and OC showed up.

**Comment by dakanji** [comment]
> I said my rEFInd was in an EFI partition on a USB stick and OC showed up.

You actually said `rEFInd is on a USB inside the ESP` which is caused the confusion. Thanks for clearing this up.

It could be that your finding is related to the disk format perhaps. The code may only be targetting certain formats as the comments from the rEFInd developer I posted in my first response stated. 

Either way, how I explained is how it operates for me and apparently for the OP as mentioned earlier. 

It appears this has turned into a drifting discussion. If there is anything substantive related to the functionality, then best to either make a new PR or continue as a discussion.

**Comment by dakanji** [comment]
Implemented but still requires allow_duplicates token.

**Pull request merged**
